src/export/export_format_spaten.cpp

```diff
@@ -0,0 +1,249 @@
+    m_fd(osmium::io::detail::open_for_writing(output_filename, overwrite)),
+    m_fsync(fsync) {
+    write_file_header();
+    reserve_block_header_space();
+    m_buffer.reserve(initial_buffer_size);
+}
+
+void ExportFormatSpaten::write_file_header() {
+    std::string fh{"SPAT"};
+    fh.append(version);
+    osmium::io::detail::reliable_write(m_fd, fh.data(), fh.size());
+}
+
+void ExportFormatSpaten::reserve_block_header_space() {
+    m_buffer.resize(m_buffer.size() + block_header_size);
+}
+
+void ExportFormatSpaten::start_feature(spaten_pbf::Geom gt, osmium::object_id_type id) {
+    m_spaten_feature.add_enum(spaten_pbf::Feature::optional_Geom_geomtype, gt);
+    m_spaten_feature.add_enum(spaten_pbf::Feature::optional_GeomSerial_geomserial, spaten_pbf::GeomSerial::wkb);
+
+    if (options().unique_id == unique_id_type::counter) {
+        std::string tagbuf;
+        protozero::pbf_builder<spaten_pbf::Tag> ptag{tagbuf};
+
+        ptag.add_string(spaten_pbf::Tag::optional_string_key, unique_id_field);
+        ptag.add_string(spaten_pbf::Tag::optional_string_value, uint64_buf(m_count));
+        ptag.add_enum(spaten_pbf::Tag::optional_ValueType_type, spaten_pbf::TagValueType::uint64);
+        m_spaten_feature.add_message(spaten_pbf::Feature::optional_Tag_tags, tagbuf);
+    } else if (options().unique_id == unique_id_type::type_id) {
+        std::string tagbuf;
+        protozero::pbf_builder<spaten_pbf::Tag> ptag{tagbuf};
+
+        char prefix = '\0';
+        if (gt == spaten_pbf::Geom::gt_node) {
+            prefix = 'n';
+        } else if (gt == spaten_pbf::Geom::gt_line) {
+            prefix = 'w';
+        } else if (gt == spaten_pbf::Geom::gt_poly) {
+            prefix = 'a';
+        }
+
+        ptag.add_string(spaten_pbf::Tag::optional_string_key, unique_id_field);
+        ptag.add_string(spaten_pbf::Tag::optional_string_value, prefix + std::to_string(id));
+        ptag.add_enum(spaten_pbf::Tag::optional_ValueType_type, spaten_pbf::TagValueType::string);
+        m_spaten_feature.add_message(spaten_pbf::Feature::optional_Tag_tags, tagbuf);
+    }
+}
+
+void ExportFormatSpaten::node(const osmium::Node& node) {
+    start_feature(spaten_pbf::Geom::gt_node, node.id());
+    m_spaten_feature.add_string(spaten_pbf::Feature::optional_string_geom, m_factory.create_point(node));
+    finish_feature(node);
+}
+
+void ExportFormatSpaten::way(const osmium::Way& way) {
+    start_feature(spaten_pbf::Geom::gt_line, way.id());
+    m_spaten_feature.add_string(spaten_pbf::Feature::optional_string_geom, m_factory.create_linestring(way));
+    finish_feature(way);
+}
+
+void ExportFormatSpaten::area(const osmium::Area& area) {
+    start_feature(spaten_pbf::Geom::gt_poly, area.id());
+    m_spaten_feature.add_string(spaten_pbf::Feature::optional_string_geom, m_factory.create_multipolygon(area));
+    finish_feature(area);
+}
+
+void ExportFormatSpaten::finish_feature(const osmium::OSMObject& object) {
+    ++m_count;
```

Comment:
Maybe move this after the following code so that no unique ids are "wasted" on feature we don't write out?

Reply:
Good point, done.